book/src/marshalling-types.md

```diff
@@ -0,0 +1,252 @@
+# Marshalling types
+
+An important part of embedding Gluon is translating non-primitive types from Gluon types to
+Rust types and vice versa, allowing you to seamlessly implement rich APIs with complex types.
+This translation is called _marshalling_.
+
+## Required traits
+
+Gluon provides several traits for safely marshalling types to and from Gluon code:
+
+- [VmType][] provides a mapping between Rust and Gluon types. It specifies the Gluon type
+  the implementing Rust type represents. All types that want to cross the Gluon/Rust boundary
+  must implement this trait.
+
+- [Getable][]: Types that implement `Getable` can be marshalled from Gluon to Rust. This
+  means you can use these types anywhere you are receiving values from Gluon, for example
+  as parameters for a function.
+
+- [Pushable][] is the counterpart to `Getable`. It allows implementing types to be marshalled
+  to Gluon. Values of these types can returned from embedded rust functions.
+
+- [Userdata][] allows a Rust type to be marshalled as completely opaque type. The Gluon code
+  will be able to receive and pass values of this type, but cannot inspect it at all. This is
+  useful for passing handle-like values, that will be mostly used by the Rust code. `Getable`
+  and `Pushable` are automatically implemented for all types that implement `Userdata`.
+
+Gluon already provides implementations for the primitive and common standard library types.
+
+## Implementing the marshalling traits for your types
+
+You can implement all of the above traits by hand, but for most cases you can also use the
+derive macros in [gluon_codegen][].
+
+You will also have to register the correct Gluon type. If you are marshalling `Userdata`, you
+can use `Thread::register_type`, otherwise you will need to provide the complete type definition
+in Gluon. When using the `serialization` feature, you can automatically generate the source
+code using the `api::typ::make_source` function.
+
+### Using derive macros
+
+Add the `gluon_codegen` crate to your `Cargo.toml` and import it:
+
+```rust,ignore
+#[macro_use]
+extern crate gluon_codegen;
+```
+
+`VmType`, `Getable` and `Pushable` can be derived for types that consist only of types that
+already implement the respective traits. In the case of `VmType` you also have to specify
+the Gluon type the Rust type maps to, using the `#[gluon(vm_type = "<gluon_type>")]` attribute.
+
+`Userdata` can be derived for any type as long as it is `Debug + Send + Sync` and has a `'static`
+lifetime.
+
+### Implementing by hand
+
+The following examples will all assume a simple struct `User<T>`, which is defined in a different
+crate. To implement the marshalling traits, we to have create a wrapper and implement the traits
+for it.
```

Comment:
Since the examples were added to `examples/marshalling.rs` it would useful to have a link to them so that readers can lookup the complete implementation!

Reply:
Yup, makes sense!